src/Control/Monad/Logic/Sequence/Internal.hs

```diff
@@ -207,27 +209,34 @@ altView (toView -> m) n = m >>= \x -> case x of
   Empty -> toView n
   h :< t -> return (h :< cat t n)
     where cat (SeqT l) (SeqT r) = SeqT (l S.>< r)
-{-# INLINE altView #-}
+{-# INLINEABLE altView #-}
 
 instance Monad m => Monad (SeqT m) where
-  {-# INLINE return #-}
+  {-# INLINEABLE return #-}
   {-# INLINEABLE (>>=) #-}
-  {-# SPECIALIZE INLINE (>>=) :: Seq a -> (a -> Seq b) -> Seq b #-}
   return = fromView . single
-  (toView -> m) >>= f = fromView $ m >>= \x -> case x of
-    Empty -> return Empty
-    h :< t -> f h `altView` (t >>= f)
-
+  (>>=) = bind
   {-# INLINEABLE (>>) #-}
-  (toView -> m) >> n = fromView $ m >>= \x -> case x of
-    Empty -> return Empty
-    _ :< t -> n `altView` (t >> n)
-
+  (>>) = then_
 #if !MIN_VERSION_base(4,13,0)
   {-# INLINEABLE fail #-}
   fail = Fail.fail
 #endif
 
+bind :: Monad m => SeqT m a -> (a -> SeqT m b) -> SeqT m b
+bind = go where
```

Comment:
Don't you mean this?

```haskell
bind m0 f = go m0
  where
    ....
```

The way you have it doesn't enable inlining.

Reply:
In my testing, if `bind` does get inlined (`INLINE` or `-O2`/`-O3`), then the performance is slightly worse. So the benefit of `INLINEABLE` (which definitely improves performance) appears to be in GHC specializing the unfolding but not inlining it.